Adding wavedrom images of the encoding

christian-herber-nxp opened the pull request.

christian-herber-nxp committed:
Adding wavedrom images of the encoding

tovine commented on images/wavedrom/c-load_store_pair.adoc:
```diff
@@ -0,0 +1,46 @@
+....
+{reg: [
+  {bits: 2, name: 'op',     type: 8, attr: ['2','C2']},
+  {bits: 5, name: 'imm',    type: 5, attr: ['5','offset[4:2|7:6]']},
```
Shouldn't the offsets here be scaled up one bit? So `offset[8:3]` instead of `[7:2]`

tovine replied:
Same applies to `C.SDSP`

christian-herber-nxp replied:
```suggestion
  {bits: 5, name: 'imm',    type: 5, attr: ['5','offset[4:3|8:6]']},
```

tovine commented on images/wavedrom/c-load_store_pair.adoc:
```diff
@@ -0,0 +1,46 @@
+  {bits: 3, name: 'rd`',     type: 3, attr: ['3', 'dest, dest[0]=0']},
+  {bits: 2, name: 'imm',     type: 2, attr: ['2', 'offset[2|6]']},
+  {bits: 3, name: 'rs1`',    type: 2, attr: ['3', 'base']},
+  {bits: 3, name: 'imm',     type: 3, attr: ['3', 'offset[5:3]']},
```
Same here, offsets `[7:3]` instead of `[6:2]` - it seems you've copied the ones for `C.LW` instead of `C.LD`?

tovine replied:
Same applies to `C.SD`

christian-herber-nxp replied:
done thanks

tovine reviewed (changes requested):
Good to have these added, do we also have a pipeline to build it and see a rendered version?

christian-herber-nxp commented:
> Good to have these added, do we also have a pipeline to build it and see a rendered version?

Nope. I do think this is needed however, and i had sent a mail to riscv-help earlier today

christian-herber-nxp commented:
> > Good to have these added, do we also have a pipeline to build it and see a rendered version?
> 
> Nope. I do think this is needed however, and i had sent a mail to riscv-help earlier today

for now, best thing during review is to build locally or to pase into https://wavedrom.com/editor.html

christian-herber-nxp commented on images/wavedrom/c-load_store_pair.adoc:
```diff
@@ -0,0 +1,46 @@
+{reg: [
+  {bits: 2, name: 'op',     type: 8, attr: ['2','C2']},
+  {bits: 5, name: 'rs2',    type: 4, attr: ['5','src, src[0]=0']},
+  {bits: 6, name: 'imm',    type: 3, attr: ['6','offset[5:2|7:6]']},
```
```suggestion
  {bits: 6, name: 'imm',    type: 3, attr: ['6','offset[5:3|8:6]']},
```

christian-herber-nxp commented on images/wavedrom/c-load_store_pair.adoc:
```diff
@@ -0,0 +1,46 @@
+{reg: [
+  {bits: 2, name: 'op',      type: 8, attr: ['2', 'C0']},
+  {bits: 3, name: 'rd`',     type: 3, attr: ['3', 'dest, dest[0]=0']},
+  {bits: 2, name: 'imm',     type: 2, attr: ['2', 'offset[2|6]']},
```
```suggestion
  {bits: 2, name: 'imm',     type: 2, attr: ['2', 'offset[7:6]']},
```

christian-herber-nxp commented on images/wavedrom/c-load_store_pair.adoc:
```diff
@@ -0,0 +1,46 @@
+{reg: [
+  {bits: 2, name: 'op',     type: 8, attr: ['2', 'C0']},
+  {bits: 3, name: 'rs2`',   type: 3, attr: ['3', 'src, src[0]=0']},
+  {bits: 2, name: 'imm',    type: 2, attr: ['2', 'offset[2|6]']},
```
```suggestion
  {bits: 2, name: 'imm',    type: 2, attr: ['2', 'offset[7:6]']},
```

christian-herber-nxp committed:
Fixing offset immediates

Signed-off-by: Christian Herber <christian.herber@oss.nxp.com>

tovine reviewed (approved):
Looks good!

wmat commented:
Note that if you are building locally, be sure to do git submodule update --remote --init to pull in the submodule code or the local build will fail to find the theme and fonts.

As for the wavedrom, if you add --verbose to the Makefile and build locally, it spits out these errors:

asciidoctor: DEBUG: images/wavedrom/load_store_pair.adoc: line 4: unknown style for literal block: wavedrom
asciidoctor: DEBUG: images/wavedrom/load_store_pair.adoc: line 15: unknown style for literal block: wavedrom
asciidoctor: DEBUG: images/wavedrom/c-load_store_pair.adoc: line 4: unknown style for literal block: wavedrom
asciidoctor: DEBUG: images/wavedrom/c-load_store_pair.adoc: line 15: unknown style for literal block: wavedrom
asciidoctor: DEBUG: images/wavedrom/c-load_store_pair.adoc: line 25: unknown style for literal block: wavedrom
asciidoctor: DEBUG: images/wavedrom/c-load_store_pair.adoc: line 37: unknown style for literal block: wavedrom

I'm working on figuring out why.

christian-herber-nxp committed:
Updating colors of compressed instruction encodings

wmat committed:
Fixed Makefile, updated docs-resources, fixed header.

Added missing required extensions to the Makefile.
Pointed docs-resources at latest version.
Fixed heading levels in header.adoc.

christian-herber-nxp committed:
Revert adding asciidoc-mathematical to requirements as it is not needed and complicated the build setup

christian-herber-nxp committed:
Adding dedicated instructions section
- single page per instruction
- just moving information, no content change

christian-herber-nxp commented:
I have done a rework of the spec to get it closer to what current specifications look like. Previously, it was in the style of RV64 & RVC, as this is where the encodings come from. Please have a look. 
[riscv-zilsd-v0.0.0.pdf](https://github.com/user-attachments/files/15534319/riscv-zilsd-v0.0.0.pdf)
Built pdf attached for those who cannot reproduce easily.

tariqkurd-repo commented on zilsd.adoc:
```diff
@@ -78,4 +118,182 @@ An implementation may have a requirement to issue a load/store pair instruction
+  {bits: 5, name: 'imm',    type: 3, attr: ['5','offset[4:3|8:6]']},
+  {bits: 5, name: 'rd',     type: 2, attr: ['5','dest≠0, dest[0]=0']},
+  {bits: 1, name: 'imm',    type: 3, attr: ['1','offset[5]']},
+  {bits: 3, name: 'funct3', type: 8, attr: ['3','C.LDSP']},
```
the value of this funct3 isn't specified anywhere, and similarly for the other instructions.
and the ≠ character doesn't render in the pdf.
I really think the encoding page needs to specify every bit in the binary encoding, including things like width=D, otherwise you have to look in another doc to see what the encoding is (or riscv-opcodes)

christian-herber-nxp replied:
i have updated the encoding to explicit in a025144.
I have no idea why the ≠ character does not render. 
It might be my local setup, because it is done in the same way for C and there it renders.

wmat replied:
The not-equal sign was not rendering because the sans-serif font inclusion was commented out in the theme.  I've updated docs-resources to fix this.  I'll open an issue to update the docs-resources pointer.

christian-herber-nxp replied:
thanks

christian-herber-nxp committed:
Making encoding of instructions explicit

tariqkurd-repo commented on zilsd.adoc:
```diff
@@ -1,18 +1,37 @@
-The Zilsd extension provides load/store pair instructions for RV32 analogous to the existing RV64 doubleword load/store instructions. The existing RV64 encodings are reused, which are reserved in RV32 without this extension.
+== Load/Store pair
+
+The Zilsd & Zcmlsd extensions provide load/store pair instructions for RV32 analogous to the existing RV64 doubleword load/store instructions. The existing RV64 encodings are reused, which are reserved in RV32 without this extension.
```
not true - the 32-bit encodings are reserved for RV32, but the 16-bit encodings replace Zcf

christian-herber-nxp replied:
right. need to get this correct without overcomplicating it.

christian-herber-nxp replied:
fixed in 04063d5

christian-herber-nxp committed:
Rewording to fix incorrect statement
- Not all instructions are generally reserved in RV32
- Zcf has an overlap with Zcmlsd in RV32

christian-herber-nxp commented:
I hope this has had sufficient review. Thanks for the work on this. If you still spot an issue, please flag in a new issue.
